Sprint 6 solution

shepilovaElena opened the pull request.

shepilovaElena committed:
тз - 6.
feat : история просмотров неограниченного размера без дублирования просмотров

shepilovaElena committed:
тз - 6.
файл для api тестов

shepilovaElena committed:
тз - 6.
файл для api тестов_

shepilovaElena committed:
тз - 6.
файл для api тестов_

shepilovaElena committed:
тз - 6.
правки метода addTaskInHistory

shepilovaElena committed:
тз - 6.
правки тестов InMemoryTaskManagerTest

shepilovaElena committed:
тз - 6.
правки тестов InMemoryTaskManagerTest

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/HistoryManager.java:
```diff
@@ -0,0 +1,14 @@
+
+   void addTaskInHistory(Task task);
+
+   void removeNode(Node<Task> node);
```
по заданию метод должен принимать id задачи, которую надо удалить, а не ноду. Это важно поскольку метод должен быть вызван из класса InMemoryTaskManager, в котором ничего не известно про ноды

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryHistoryTaskManager.java:
```diff
@@ -0,0 +1,80 @@
+import java.util.*;
+
+public class InMemoryHistoryTaskManager implements HistoryManager {
+    private List<Task> publicHistory;
```
эта переменная нужна только для метода getHistory(), поэтому ее можно целиком объявить только там

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryHistoryTaskManager.java:
```diff
@@ -0,0 +1,80 @@
+    private Map<Integer, Node<Task>> history = new HashMap<>();
+    private Node<Task> head;
+    private Node<Task> tail;
+    private Node<Task> newView;
```
аналогично про объявление в методе addTaskInHistory(). Скажу даже больше, объявление их здесь немного путает в начале

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryHistoryTaskManager.java:
```diff
@@ -0,0 +1,80 @@
+    @Override
+    public List<Task> getHistory() {
+        publicHistory = new ArrayList<>();
+        publicHistory.add(head.data);
```
если это действие убрать, то можно будет на 21 строчке в список добавлять задачу текущей ноды, а не следующей. Честно, такая архитектура метода прям глаз режет

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryHistoryTaskManager.java:
```diff
@@ -0,0 +1,80 @@
+            tail = new Node<>(task);
+            head.next = tail;
+            tail.prev = head;
+            tail.next = null;
```
это лишнее действие, у tail ссылка на next итак пока null

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryHistoryTaskManager.java:
```diff
@@ -0,0 +1,80 @@
+        if (node == head) {
+            node.next.prev = null;
+            head = node.next;
+            node.next = null;
```
это лишнее действие, на выходе из метода про node все соседи забудут, поэтому не важно на что она там ссылается

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryHistoryTaskManager.java:
```diff
@@ -0,0 +1,80 @@
+            node.next.prev = null;
+            head = node.next;
+            node.next = null;
+            history.put(head.data.getId(), head);
```
это действие не нужно, такая нода итак есть в history

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryHistoryTaskManager.java:
```diff
@@ -0,0 +1,80 @@
+        } else if (node == tail) {
+            node.prev.next = null;
+            tail = node.prev;
+            node.prev = null;
```
это тоже лишнее действие, потому что про node все забудут

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryHistoryTaskManager.java:
```diff
@@ -0,0 +1,80 @@
+            node.prev.next = null;
+            tail = node.prev;
+            node.prev = null;
+            history.put(node.data.getId(), tail);
```
и это лишнее, потому что в history уже все есть

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryHistoryTaskManager.java:
```diff
@@ -0,0 +1,80 @@
+        } else {
+            node.next.prev = node.prev;
+            node.prev.next = node.next;
+            node.next = null;
```
эти два последних действия тоже лишние

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -115,18 +115,18 @@ private void changeEpicStatus(Epic epic) {
     // удаление всех задач
     @Override
-    public void deleteAllTasks(){
+    public void deleteAllTasks() {
```
нужно еще удалять из истории просмотров

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -115,18 +115,18 @@ private void changeEpicStatus(Epic epic) {
 
     @Override
-    public void deleteAllEpics(){
+    public void deleteAllEpics() {
```
аналогично

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -115,18 +115,18 @@ private void changeEpicStatus(Epic epic) {
 
     @Override
-    public void deleteAllSubtasks(){
+    public void deleteAllSubtasks() {
```
аналогично

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -206,14 +206,14 @@ public ArrayList<Subtask> getAllSubtasks() {
     // удаление по id
     @Override
-    public void deleteTaskById(int id){
+    public void deleteTaskById(int id) {
```
аналогично

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -206,14 +206,14 @@ public ArrayList<Subtask> getAllSubtasks() {
     @Override
-    public void deleteEpicById(int id){
-        if(epics.containsKey(id)) {
+    public void deleteEpicById(int id) {
```
аналогично

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryTaskManager.java:
```diff
@@ -222,7 +222,7 @@ public void deleteEpicById(int id){
     @Override
     public void deleteSubtaskById(int id) {
-        if(epics.containsKey(id)) {
+        if (epics.containsKey(id)) {
```
аналогично

shepilovaElena committed:
тз - 6.
добавление метода removeTaskFromHistory
соответствующее исправление методов удаления задач
тест removeTaskFromHistoryTest

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/HistoryManager.java:
```diff
@@ -0,0 +1,16 @@
+
+   void addTaskInHistory(Task task);
+
+   void removeNode(Node<Task> node);
```
Теперь этот метод можно убрать из интерфейса, поскольку он должен быть внутренним методом класса InMemoryHistoryTaskManager

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryHistoryTaskManager.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @Override
+    public void removeNode(Node<Task> node) {
```
метод используется только внутри класса => должен быть private

irek-valiev commented on src/ru/yandex/practicum/javakanban/manager/InMemoryHistoryTaskManager.java:
```diff
@@ -0,0 +1,83 @@
+    @Override
+    public void removeNode(Node<Task> node) {
+        if (history.size() == 1) {
+            node.next = null;
```
вот этой логики я вообще не понял, если удаляется единственная нода, то в ней итак node.next уже нулл? И самое главное, никакого удаления ноды здесь не происходит. Нужно обнулить head

shepilovaElena committed:
тз - 6.
правки метода removeNode
